### PR TITLE
Certs on Manifold

### PR DESCRIPTION
**Pull request by akrolsmir**
See https://manifoldmarkets.notion.site/Certs-4319707df7be43489b0f0620ed095511

### CODE REVIEW

**Comment by akrolsmir** [comment]
Mostly looking for a review on the way the data is structured here, through a log of Txns rather than Bets; and secondarily a review of the Uniswapv2 mechanism -- thanks @mantikoros @mqp!

The UI will continue to change a lot, so don't worry too much about it

File: common/calculate/cert.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  // 3) Pay out that much M$ to each holder (assume that all pool shares belong to cert creator)
+  const ownership = getCertOwnershipUsers(providerId, txns)
+  const totalShares = sum(Object.values(ownership))
+  const dividendPerShare = totalDividend / totalShares
+  const payouts = Object.entries(ownership).map(([ownerId, shares]) => ({
+    userId: ownerId,
+    payout:
+      shares * dividendPerShare -
+      // Set a negative total payout for the provider
+      (ownerId === providerId ? totalDividend : 0),
```
Comment (mantikoros) [inline common/calculate/cert.ts:57]: What if the owner and dividend payer are different people? Might be easier to deduct dividend payment in a different transaction.

File: common/calculate/uniswap2.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+export function afterSwap(
+  pool: { [outcome: string]: number },
+  token: 'M$' | 'SHARE',
+  amount: number
+) {
+  const k = pool['M$'] * pool['SHARE']
+  const other = token === 'M$' ? 'SHARE' : 'M$'
+  return {
+    [token]: pool[token] + amount,
+    [other]: k / (pool[token] + amount),
```
Comment (mantikoros) [inline common/calculate/uniswap2.ts:26]: Check for overflow/underflow if pool[token] + amount is close to 0.

File: common/contract.ts [hunk 3]
```diff
@@ -107,6 +111,21 @@ export type CPMM = {
   }
 }
 
+export type Uniswap2 = {
+  mechanism: 'uniswap-2'
+  // outcome can be e.g. 'M$' or a 'SHARE'
+  pool: { [outcome: string]: number }
```
Comment (mantikoros) [inline common/contract.ts:117]: Why not just use fields for shares and mana? Multivariable case probably is not very common.

File: common/contract.ts [hunk 4]
```diff
@@ -107,6 +111,21 @@ export type CPMM = {
...
+  // Unused; just for making Typescript happy, because code paths
+  // assume this thing falls under DPM-2
```
Comment (mantikoros) [inline common/contract.ts:119]: There must be a way to get typescript to not complain. Adding an extra field is bad.

**Review by mantikoros — APPROVED** [review]
Always check for pool overflow with Uniswap (ie make sure each item in the pool stays above some epsilon value), but otherwise LGTM.

File: common/contract.ts (hunk 4 above)
Comment (akrolsmir) [reply common/contract.ts:119]: Yeah, will look into getting rid of the field!

File: common/contract.ts (hunk 3 above)
Comment (akrolsmir) [reply common/contract.ts:117]: Wanted to leave open the mechanism in the future to eg swap two different shares/coins, I think? But maybe YAGNI. I don't think it matters much either way, here.

**Pull request merged**